---
title: "HTML Report"
weight: 15
pre: '<i class="fas fa-chart-bar"></i> '
plotly: true
---
### Concept

---
Alignments generated by `Captus` can be directly input to phylogenetic tree reconstruction tools such as [`IQ-TREE`](http://www.iqtree.org).
However, before doing so, it is common practice to **filter out noisy alignments** that contain only a few samples or excess of gaps.
This little hassle allows to obtain **more accurate phylogenetic inference with less computation time**.
In addition, since `Captus` generates several versions of alignments with different formats, trimming conditions, paralog filtering conditions, and so on, you may be wondering which version to use for phylogenetic inference.

Open the report, `captus-assembly_align.report.html` in your browser (Microsoft Edge, Google Chrome, Mozilla Firefox, Safari, etc., internet connection required) to see which alignments should be filtered out, and to compare general statistics for each version of alignments!
{{% notice info %}}

- All original data for the report is stored in `captus-assembly_align.stats.tsv`.
- Since all tables in the report are created using [`Plotly`](https://plotly.com/python), you can use some useful functions such as rearrange the column order by dragging and download plot as a PNG.
For more information, please visit the following site:

  - <https://plotly.com/chart-studio-help/getting-to-know-the-plotly-modebar>
{{% /notice %}}

### Example

---
Here is an example of the report for the [`Angiosperms353`](https://github.com/mossmatters/Angiosperms353) loci extracted from four hybridization-capture samples.  
This table shows the general statistics of the different versions of alignment for each locus.  
Each version of the alignments differs in the combination of whether it is trimmed or untrimmed, which paralog filter was applied, and whether it was aligned with the reference sequences or not.
{{< plotly json="/plotly/alignment_report.json" height="500px" >}}
{{% notice note %}}

- If your results contain multiple marker types, the report will include a separete table for each marker type as well as a global table for all marker types.
- The cells are colored according to their relative value within each row (green = high; pink = low).
- The table rows are sorted in descending order by the mean value of each row.
{{% /notice %}}
{{% expand "Description of each column" %}}
|Column|Description|
|-|-|
|**Marker type**|Marker type (This column only appear on a global table for all marker types)<br>For a description of each format, see [`-k, --markers`](http://localhost:1313/assembly/align/options/#-k---markers) option|
|**Format**|Format of the alignment<br>For a description of each format, see [`-f, --formats`](http://localhost:1313/assembly/align/options/#-f---formats) option|
|**Locus**|Locus/marker name|

The subsequent columns correspond to the different versions of the alignments in your results.  
You will have up to 12 columns as follows:
|Column|Trimming|Paralog filter|With reference|Location|
|-|-|-|-|-|
|**02_aligned_untrimmed<br>└01_unfiltered**|No|No|**Yes**|`02_aligned_untrimmed/01_unfiltered`|
|**02_aligned_untrimmed<br>└02_fast**|No|**Fast**|**Yes**|`02_aligned_untrimmed/02_fast`|
|**02_aligned_untrimmed<br>└03_careful**|No|**Careful**|**Yes**|`02_aligned_untrimmed/03_careful`|
|**02_aligned_untrimmed<br>└01_unfiltered_no_refs**|No|No|No|`02_aligned_untrimmed/01_unfiltered_no_refs`|
|**02_aligned_untrimmed<br>└02_fast_no_refs**|No|**Fast**|No|`02_aligned_untrimmed/02_fast_no_refs`|
|**02_aligned_untrimmed<br>└03_careful_no_refs**|No|**Careful**|No|`02_aligned_untrimmed/03_careful_no_refs`|
|**03_aligned_trimmed<br>└01_unfiltered**|**Yes**|No|**Yes**|`03_aligned_trimmed/01_unfiltered`|
|**03_aligned_trimmed<br>└02_fast**|**Yes**|**Fast**|**Yes**|`03_aligned_trimmed/02_fast`|
|**03_aligned_trimmed<br>└03_careful**|**Yes**|**Careful**|**Yes**|`03_aligned_trimmed/03_careful`|
|**03_aligned_trimmed<br>└01_unfiltered_no_refs**|**Yes**|No|No|`03_aligned_trimmed/01_unfiltered_no_refs`|
|**03_aligned_trimmed<br>└02_fast_no_refs**|**Yes**|**Fast**|No|`03_aligned_trimmed/02_fast_no_refs`|
|**03_aligned_trimmed<br>└03_careful_no_refs**|**Yes**|**Careful**|No|`03_aligned_trimmed/03_careful_no_refs`|

{{% /expand %}}

### Features

---

#### Format dropdown

By default, the tables contain the information about all formats included in your results.
This dropdown allows you to filter each table and extract information about one specific format.
(Note that when you switch the dropdown, the table will be automatically re-sorted and re-colored according to the values.)

For a description of each format, see [`-f, --formats`](http://localhost:1313/assembly/align/options/#-f---formats) option.

---

#### Variable dropdown

By default, the tables show the number of sequences in each alignment.
This dropdown allows you to switch the variable to be shown.
(Note that when you switch the dropdown, the table will be automatically re-sorted and re-colored according to the values.)

There are eight options:
|Variable|Description|
|-|-|
|**Number of sequences** (default)|Number of sequences in the alignment|
|**Alignment length**|Length of the alignment in base pairs (for `NT`, `GE`, `GF`, `MA`, `MF` formats) or amino acids (for `AA` format)|
|**Informaive sites**|Number of parsimony-informative sites that have at least two different (unambiguous) characters and of those at least two should occur in at least two samples|
|**Constant sites**|Number of constant sites that have only a single character over all sequences, or sites having only one unambiguous character plus ambiguous characters which include that unambiguous character|
|**Singleton sites**|Number of singleton sites that have at least two types of characters with at most one unambiguous character occuring multiple times while all other characters occur only once|
|**Patterns**|Number of sites that have different patterns of characters|
|**Mean pairwise identity (%)**|Percentage of pairwise characters that are identical in the alignment (including gap versus non-gap residues, but excluding gap versus gap residues)|
|**Missingness (%)**|Percentage of gaps and Ns or Xs in the alignment|

For more information about sites and patterns, read [<i class="fab fa-readme"></i> IQ-TREE documentation](http://www.iqtree.org/doc/Frequently-Asked-Questions#how-does-iq-tree-treat-identical-sequences)

---
Created by [Gentaro Shigita]({{< ref "../../credits/#gentaro-shigita">}}) (11.08.2021)  
Last modified by [Gentaro Shigita]({{< ref "../../credits/#gentaro-shigita">}}) (29.09.2021)
